firebase-firestore/src/test/java/com/google/firebase/firestore/remote/BloomFilterTest.java

```diff
@@ -0,0 +1,252 @@
+  }
+
+  @Test
+  public void mightContainCanProcessNonStandardCharacters() {
+    // A non-empty BloomFilter object with 1 insertion : "ÀÒ∑"
+    BloomFilter bloomFilter = new BloomFilter(new byte[] {(byte) 237, 5}, 5, 8);
+    assertTrue(bloomFilter.mightContain("ÀÒ∑"));
+    assertFalse(bloomFilter.mightContain("Ò∑À"));
+  }
+
+  @Test
+  public void mightContainOnEmptyBloomFilterShouldReturnFalse() {
+    BloomFilter bloomFilter = new BloomFilter(new byte[0], 0, 0);
+    assertFalse(bloomFilter.mightContain("a"));
+  }
+
+  @Test
+  public void mightContainWithEmptyStringShouldReturnFalse() {
+    BloomFilter emptyBloomFilter = new BloomFilter(new byte[0], 0, 0);
+    assertFalse(emptyBloomFilter.mightContain(""));
+    BloomFilter nonEmptyBloomFilter = new BloomFilter(new byte[] {(byte) 255}, 0, 1);
+    assertFalse(nonEmptyBloomFilter.mightContain(""));
+  }
+
+  @Test
+  public void bloomFilterToString() {
+    BloomFilter emptyBloomFilter = new BloomFilter(new byte[0], 0, 0);
+    assertEquals(emptyBloomFilter.toString(), "BloomFilter{hashCount=0, size=0, bitmap=\"\"}");
+    BloomFilter nonEmptyBloomFilter = new BloomFilter(new byte[] {1}, 1, 1);
+    assertEquals(
+        nonEmptyBloomFilter.toString(), "BloomFilter{hashCount=1, size=7, bitmap=\"AQ==\"}");
+  }
+
+  /**
+   * Golden tests are generated by backend based on inserting n number of document paths into a
+   * bloom filter.
+   *
+   * <p>Full document path is generated by concatenating documentPrefix and number n, eg,
+   * projects/project-1/databases/database-1/documents/coll/doc12.
+   *
+   * <p>The test result is generated by checking the membership of documents from documentPrefix+0
+   * to documentPrefix+2n. The membership results from 0 to n is expected to be true, and the
+   * membership results from n to 2n is expected to be false with some false positive results.
+   */
+  private void runGoldenTest(String testFile) throws Exception {
+    String resultFile = testFile.replace("bloom_filter_proto", "membership_test_result");
+
+    JSONObject testJson = readJsonFile(testFile);
+    JSONObject resultJSON = readJsonFile(resultFile);
+
+    JSONObject bits = testJson.getJSONObject("bits");
+    String bitmap = bits.getString("bitmap");
+    int padding = bits.getInt("padding");
+    int hashCount = testJson.getInt("hashCount");
+    BloomFilter bloomFilter =
+        new BloomFilter(Base64.getDecoder().decode(bitmap), padding, hashCount);
+
+    String membershipTestResults = resultJSON.getString("membershipTestResults");
+
+    // Run and compare mightContain result with the expectation.
+    for (int i = 0; i < membershipTestResults.length(); i++) {
+      boolean expectedMembershipResult = membershipTestResults.charAt(i) == '1';
+      boolean mightContain = bloomFilter.mightContain(GOLDEN_DOCUMENT_PREFIX + i);
+      assertEquals(
+          "mightContain() result doesn't match the expectation. File: "
```

Comment:
Consider this re-wording of the failure message: `mightContain(...) returned X, but expected Y`.